### ISSUE TITLE
OnAnyFailure throw custom exception

### ISSUE THREAD

**Issue by tarastriska** [JeremySkinner/FluentValidation#973]
Hi,

In my validator I have smth like this

`RuleFor(x => x.MyProperty).NotEmpty().Length(1, 100).OnAnyFailure(obj => throw new MyCustomException(errorCode));`

Then in controller I call validate method. I expect exception to be thrown but code continues to execute.
Am I misusing it?

Note: It's .NET Core 2.1

**Comment by JeremySkinner** [comment]
That's not really the intention of `OnFanyFailure` - if you want to throw an exception you should use the call the `ValidateAndThrow` method on the root validator. That being said, it should work just fine, so if you're not seeing an exception being thrown it probably means validation is succeeding.

**Comment by tarastriska** [comment]
Ok, sorry for confusion. It works.
I need specific error due to some logic.
But now I have another question.
Can I somehow get an error message in OnAnyFailure callback?

**Comment by JeremySkinner** [comment]
No, the `OnAnyFailure` callback only gives you a reference to the object currently being validated. Perhaps you could provide some more details on what you're trying to accomplish, and then I might be able to suggest a better way...

**Comment by tarastriska** [comment]
Ok,

The scenario is simple. I wanna validate each property with rule chains. If any validation for property fails I need to raise custom exception with 2 properties like `code:SomePropertyIsInvalid and message:Property is empty`. So it would be nice to have ability to get value provided in `WithMessage `in scope of `OnAnyFailure`

**Comment by JeremySkinner** [comment]
So you can do this already with `OnFailure`:

```csharp
RuleFor(x => x.MyProperty).NotEmpty().Length(1, 100).OnFailure((obj, context, errorMessage) => ...);
```

But note that `OnFailure` only applies to the *validator that immediately precedes it*, so it knows which validator to receive the message from. So in my above example, `OnFailure` would be invoked if the `Length` validator fails, but not if the `NotEmpty` fails. So if you wanted this to happen for both, you'd need 2 `OnFailure` calls:

```csharp
RuleFor(x => x.MyProperty)
   .NotEmpty().OnFailure((obj, context, errorMessage) => ...)
  .Length(1, 100).OnFailure((obj, context, errorMessage) => ...);
```

The reason that the message is not available in `OnAnyFailure` is becauase this is executed after all the validators in the chain have run - it only indicates that there was at least one failure and doesn't know which validator (or validators if more than one failed) caused the failure. 

We could look at adding a new overload of `OnAnyFailure` in the future that would receive a collection of error messages. Would that be helpful?

**Comment by tarastriska** [comment]
Thanks a lot. I guess the thread can be closed

**Comment by JeremySkinner** [comment]
I'll keep it open to track development of the `OnAnyFailure` change.

**Comment by JeremySkinner** [comment]
Implemented in ffd123b which will be part of 9.0.
